device/protocol_mqtt_v1.go

```diff
@@ -270,13 +286,69 @@ func (d *Device) sendMqttV1MessageInternal(astarteInterface interfaces.AstarteIn
 		return err
 	}
 
-	topic := fmt.Sprintf("%s/%s%s", d.getBaseTopic(), astarteInterface.Name, interfacePath)
-	// TODO: Handle this token
-	_ = d.m.Publish(topic, qos, false, doc)
+	// if it's a property, we need to set it
+	if astarteInterface.Type == interfaces.PropertiesType {
+		go d.storeProperty(astarteInterface.Name, interfacePath, doc)
+	}
+
+	if d.isSendingStoredMessages {
+		fmt.Println("Sending previously stored messages with non-discard retention, the current message may be scheduled later")
+	}
+	message := makeAstarteMessageInfo(mapping.Expiry, mapping.Retention, astarteInterface.Name, interfacePath, qos, doc)
+	d.messageQueue <- message
+
+	return nil
+}
+
+func (d *Device) publishMessage(message astarteMessageInfo) error {
+	topic := fmt.Sprintf("%s/%s%s", d.getBaseTopic(), message.InterfaceName, message.Path)
+
+	// MQTT client returns `true` to IsConnected()
+	// even if it is actually reconnecting
+	if !d.m.IsConnectionOpen() {
+		fmt.Printf("Message on %s not delivered: MQTT client disconnected\n", topic)
+		d.storeMessage(message)
+	} else {
+
+		t := d.m.Publish(topic, message.Qos, false, message.Payload)
+		// We wait for either successful delivery or error
+		_ = t.Wait()
+		if t.Error() != nil {
+			fmt.Printf("Message on %s not delivered: %s\n", topic, t.Error().Error())
+			d.storeMessage(message)
```

Comment:
If you fail here, shouldn't you _prepend_ instead of appending (which `storeMessage` does)? Otherwise if you queue is `[1, 2, 3]` and you fail to process `1`, you're queue will become `[2, 3, 1]`, dropping the ordering guarantees

Reply:
Nvm, actually you're consuming from the channel and appending to a separate queue